### ISSUE TITLE
policy: branching

### ISSUE THREAD

**Issue by ngjoh** [nexi-intra/development#1]
Certainly! Here’s a branching strategy that can support canary, beta, preview, and production releases, with only production releases getting semantic tags:

### Branching Strategy

1. **Main Branches:**
   - `main` (or `master`): This branch should always reflect the production-ready state of your codebase. Only stable releases are merged into this branch.
   - `develop`: This branch serves as the integration branch for features and fixes. It reflects the latest delivered development changes for the next release.

2. **Release Branches:**
   - `release/canary`: This branch is used for canary releases. It allows you to test new features in a production-like environment with a limited audience.
   - `release/beta`: This branch is for beta releases. It is a more stable release compared to canary, intended for a larger group of users.
   - `release/preview`: This branch is for preview releases, which are nearly final and intended for users to see upcoming features.

3. **Feature and Bugfix Branches:**
   - `feature/<feature-name>`: These branches are used for developing new features.
   - `bugfix/<bugfix-name>`: These branches are for hotfixes and bug fixes.

### Workflow

1. **Development:**
   - Create a feature branch from `develop` for each new feature: `feature/<feature-name>`.
   - Once the feature is complete, create a pull request (PR) to merge the feature branch into `develop`.

2. **Canary Release:**
   - When ready for a canary release, merge `develop` into `release/canary`.
   - Deploy from the `release/canary` branch to the canary environment.
   - Test and gather feedback.

3. **Beta Release:**
   - Once the canary release is stable, merge `release/canary` into `release/beta`.
   - Deploy from the `release/beta` branch to the beta environment.
   - Test and gather feedback.

4. **Preview Release:**
   - When the beta release is stable, merge `release/beta` into `release/preview`.
   - Deploy from the `release/preview` branch to the preview environment.
   - Test and gather feedback.

5. **Production Release:**
   - Once the preview release is stable and approved, merge `release/preview` into `main`.
   - Create a semantic version tag (e.g., `v1.0.0`) on the `main` branch.
   - Deploy from the `main` branch to the production environment.

6. **Hotfixes:**
   - For urgent fixes in production, create a `hotfix/<hotfix-name>` branch from `main`.
   - Once the hotfix is complete, merge it back into `main` and tag it with a new semantic version.
   - Also merge the hotfix back into `develop`, and subsequently into `release/*` branches if applicable.

### Example Branch Structure

```plaintext
main
|
|-- develop
    |
    |-- feature/feature-1
    |
    |-- release/canary
        |
        |-- release/beta
            |
            |-- release/preview
```

### Semantic Versioning

- Production releases on the `main` branch should be tagged with semantic versioning tags (e.g., `v1.0.0`, `v1.1.0`, `v2.0.0`).
- Use meaningful version numbers to indicate the nature of changes (major, minor, patch).

### Tips

- Automate the merging process and deployments using CI/CD tools.
- Ensure thorough testing at each stage before moving to the next release branch.
- Use branch protection rules to enforce code review and testing before merging.

This strategy allows you to manage multiple release stages efficiently while maintaining a clear path to production.
